sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,253 @@
+    private static final Map<AsyncDocumentClient.OpenConnectionAggressivenessHint, ConcurrencyConfiguration> concurrencySettings = new HashMap<>();
+    private final IOpenConnectionsHandler openConnectionsHandler;
+    private final RntbdEndpoint.Provider endpointProvider;
+    private Disposable openConnectionBackgroundTask;
+
+    static {
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.DEFENSIVE, new ConcurrencyConfiguration(Configs.getOpenConnectionsDefensiveConcurrency(), Configs.getOpenConnectionsDefensiveConcurrency()));
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE, new ConcurrencyConfiguration(Configs.getOpenConnectionsAggressiveConcurrency(), Configs.getOpenConnectionsAggressiveConcurrency()));
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider, Duration aggressiveConnectionEstablishmentDuration) {
+        this.openConnectionsTaskSink = Sinks.many().multicast().onBackpressureBuffer();
+        this.openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer();
+        this.aggressivenessHint = new AtomicReference<>(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE);
+        this.endpointsUnderMonitorMap = new ConcurrentHashMap<>();
+        this.openConnectionsHandler = new RntbdOpenConnectionsHandler(endpointProvider);
+        this.endpointProvider = endpointProvider;
+
+        // start as a background task
+        openConnectionBackgroundTask = this.getOpenConnectionsPublisher();
+        Flux<Integer> backgroundOpenConnectionsSinkReInstantiationTask = Flux
+                .just(1)
+                .delayElements(aggressiveConnectionEstablishmentDuration);
+
+        if (aggressiveConnectionEstablishmentDuration != null && aggressiveConnectionEstablishmentDuration.compareTo(Duration.ZERO) > 0) {
+            backgroundOpenConnectionsSinkReInstantiationTask
+                    .doOnComplete(() -> reInstantiateOpenConnectionsPublisherAndSubscribe())
+                    .subscribeOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+                    .subscribe();
+        }
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider) {
+        this(endpointProvider, null);
+    }
+
+    public synchronized OpenConnectionTask submitOpenConnectionTaskOutsideLoop(
+        String collectionRid,
+        URI serviceEndpoint,
+        Uri addressUri,
+        int minConnectionsRequiredForEndpoint) {
+
+        OpenConnectionTask openConnectionTask = new OpenConnectionTask(collectionRid, serviceEndpoint, addressUri, minConnectionsRequiredForEndpoint);
+        this.submitOpenConnectionTaskOutsideLoopInternal(openConnectionTask);
+
+        return openConnectionTask;
+    }
+
+    private synchronized void submitOpenConnectionTaskOutsideLoopInternal(OpenConnectionTask openConnectionTask) {
```

Comment:
this doesn't need to be synchronized, since it is only called once and that too also within the above synchronized method.

Reply:
Resolved it offline.